sycl/include/sycl/ext/oneapi/properties/property.hpp

```diff
@@ -0,0 +1,174 @@
+//==---------- properties.hpp --- SYCL extension property tooling ----------==//
+//
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+
+// HOW-TO: Add new compile-time property
+//  1. Add a new enumerator to
+//     `sycl::ext::oneapi::experimental::detail::PropKind` representing the new
+//     property. Increment
+//     `sycl::ext::oneapi::experimental::detail::PropKind::PropKindSize`
+//  2. Define property key class with `value_t` that must be `property_value`
+//     with the first template argument being the property class itself. The
+//     name of the key class must be the property name suffixed by `_key`, i.e.
+//     for a property `foo` the class should be named `foo_key`.
+//  3. Add an `inline constexpr` variable in the same namespace as the property
+//     key. The variable should have the same type as `value_t` of the property
+//     class, e.g. for a property `foo`, there should be a definition
+//     `inline constexpr foo_key::value_t foo`.
+//  4. Specialize `sycl::ext::oneapi::experimental::is_property` and
+//     `sycl::ext::oneapi::experimental::is_property_of` for the property key
+//     class.
+//  5. Specialize `sycl::ext::oneapi::experimental::detail::PropertyToKind` for
+//     the new property key class. The specialization should have a `Kind`
+//     member with the value equal to the enumerator added in 1.
+//  6. Specialize
+//     `sycl::ext::oneapi::experimental::detail::IsCompileTimeProperty` for the
+//     new property key class. This specialization should derive from
+//     `std::true_type`.
+/******************************** EXAMPLE **************************************
+------------- sycl/include/sycl/ext/oneapi/properties/property.hpp -------------
+// (1.)
+enum PropKind : uint32_t {
+  ...
+  Bar,
+  PropKindSize = N + 1, // N was the previous value
+};
+---------------------- path/to/new/property/file.hpp ---------------------------
+namespace sycl::ext::oneapi::experimental {
+
+// (2.)
+struct bar_key {
+  using value_t = property_value<bar_key>;
+};
+
+// (3.)
+inline constexpr bar_key::value_t bar;
+
+// (4.)
+template <> struct is_property<bar_key> : std::true_type {};
+// Replace SYCL_OBJ with the SYCL object to support the property.
+template <> struct is_property_of<bar_key, SYCL_OBJ> : std::true_type {};
```

Comment:
what if the SYCL_OBJ class itself requires template specialization?  Use a template template? Proxy? Will we have order-of-evaluation issues?

Reply:
`is_property_of` is already a thing in SYCL 2020. For example, if we specialize for buffer we have something long the lines of
```c++
template <typename T, int Dimensions, typename AllocatorT>
struct is_property_key_of<bar_key, buffer<T, Dimensions, AllocatorT, void>> : std::true_type {};
```
I don't believe there should be any issues with that.